DOC: Add scaling to large datasets section

TomAugspurger opened the pull request:
Closes https://github.com/pandas-dev/pandas/issues/28315

TomAugspurger committed:
DOC: Add scaling to large datasets section

Closes https://github.com/pandas-dev/pandas/issues/28315

jreback commented on pandas/util/testing.py:
```diff
@@ -1652,6 +1652,87 @@ def makeMultiIndex(k=10, names=None, **kwargs):
+]
+
+
+def make_timeseries(start="2000-01-01", end="2000-12-31", freq="1D", seed=None):
```
i know this is how dask names things but this is confusing here as it returns a frame

also is it possible to make this as a wrapper around makeTimessries and/or getTimeSeriesData (which, which should rename to python case, OT)

jreback commented on doc/.gitignore:
```diff
@@ -0,0 +1,4 @@
+data/
+timeseries.csv
+timeseries.parquet
+timeseries_wide.parquet
```
need a newline

jreback commented on doc/source/user_guide/scale.rst:
```diff
@@ -0,0 +1,343 @@
+
+You see more dask examples at https://examples.dask.org.
+
+Use chunking
```
would move chuncking up before dask (as its a smaller section can get lost)

jreback replied:
*maybe* can mention some of the ecosystem libraries: vaex, modin, spark

TomAugspurger replied:
Originally I had the chunking earlier. I moved it down here, since I'm not sure I would ever recommend it as an option...

Perhaps I'll move it up, with the caveat that there are easier ways (Dask, etc.). And maybe show value counts, but say how difficult doing a `groupby` would be out of core.

jreback replied:
yeah was just a thought, though chunking is pretty appropriate for things like file-to-file transformations (e.g. csv -> parquet) for example

TomAugspurger committed:
emphasize memory

TomAugspurger committed:
fixups

TomAugspurger committed:
code checks

TomAugspurger committed:
bump pyarrow

TomAugspurger committed:
update deps

TomAugspurger committed:
include in user_guide index

WillAyd commented on doc/source/user_guide/scale.rst:
```diff
@@ -0,0 +1,362 @@
+
+   import pandas as pd
+   import numpy as np
+   from pandas.util.testing import make_timeseries
```
Just thinking through implications but is this something we really want to do? I feel like this is an unnecessary API exposure

TomAugspurger replied:
Open to suggestions here. I've tried to make the API exposure as small as possible. But I need a way to

1. Generate a semi-realistic dataset of arbitrary size.
2. Without distracting from the overall message.

So I think the options are what I have here, or a private method like `_make_timeseries()` and the docs just describes the raw data contained in the file on disk (but we hide the generation of that file).

WillAyd replied:
I think private method and hiding import would be preferable; maybe just a comment preceding first usage saying `# arbitrary large frame` or something to the effect. The user shouldn't care about the import machinery

TomAugspurger replied:
Just to be clear, if I make it private, I'm not going to show it being imported.

TomAugspurger replied:
Though I'll note in passing that a way to generate realistic, sample random datasets is nice :)

WillAyd commented on doc/source/user_guide/scale.rst:
```diff
@@ -0,0 +1,362 @@
+   files = list(pathlib.Path("data/timeseries/").glob("ts*.parquet"))
+   files
+
+Now we'll implement an out-of-core ``value_counts``. The peak memory usage of this
```
Does the `out-of-core` label apply here?

TomAugspurger replied:
Can you clarify?

If you're asking if this is an example of an out-of-corse algorithm, then yes. But I think you may be asking something else :)

WillAyd replied:
No I just didn't think this was an `out-of-core` algorithm since we do in fact load an entire data frame into memory (granted via iteration)

TomAugspurger replied:
It's out-of-core in the sense that we aren't loading the entire "logical dataset" on disk into memory at once. The algorithm scales to arbitrary sizes as long as each chunk (individual parquet file) fits in memory.

I'll try to clarify this in the example.

WillAyd commented:
Thanks for addressing comments. Generally looks good and I think a welcome addition to docs

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into scale

jschendel commented on doc/source/user_guide/scale.rst:
```diff
@@ -0,0 +1,362 @@
+       counts = counts.add(df['name'].value_counts(), fill_value=0)
+   counts.astype(int)
+
+Some readers, like :meth:`pandas.read_csv` offer parameters to control the
```
needs a comma after the :meth:\`pandas.read_csv\`

jschendel commented on doc/source/user_guide/scale.rst:
```diff
@@ -0,0 +1,362 @@
+.. ipython:: python
+
+   col = "timestamp"
+   %time pd.read_csv("timeseries.csv", index_col=col, parse_dates=[col])
```
Should this also have the `_ =` like what's being done with `read_parquet`?

TomAugspurger committed:
update

jschendel commented on doc/source/user_guide/scale.rst:
```diff
@@ -0,0 +1,404 @@
+Scaling to large datasets
+*************************
+
+Pandas provide data structures for in-memory analytics, which makes using pandas
```
"provide" --> "provides"

TomAugspurger commented:
And thank you for the review :)

FYI, this adds ~30s to the doc build on my machine (haven't checked the slowdown on CI, but presumably its longer). Is that unacceptable? I can make the examples smaller if needed.

jschendel reviewed (commented):
Looks good overall, just some nitpicks on my part.

TomAugspurger committed:
fixups

TomAugspurger commented:
FYI, rendered version:

![FireShot Capture 007 - Scaling to large datasets — pandas 0 25 0+418 ga7fb97f44a dirty documen_ - ](https://user-images.githubusercontent.com/1312546/65462754-0f615b80-de1c-11e9-88c1-a939e6fb8aab.png)

TomAugspurger commented:
Any other thoughts here?

WillAyd commented on doc/source/user_guide/scale.rst:
```diff
@@ -0,0 +1,404 @@
+.. ipython:: python
+
+   %%time
+   files = list(pathlib.Path("data/timeseries/").glob("ts*.parquet"))
```
I don't think necessary to encapsulate this in list

TomAugspurger replied:
Yeah, that's a leftover from before, when I printed out `files`. But that's covered by the `tree` output earlier on now that we aren't showing `make_timeseries`.

WillAyd commented on doc/source/user_guide/scale.rst:
```diff
@@ -0,0 +1,404 @@
+       df = pd.read_parquet(path)
+       # ... plus a small Series `counts`, which is updated.
+       counts = counts.add(df['name'].value_counts(), fill_value=0)
+   counts.astype(int)
```
Is this necessary? Just seems like some cruft in here for dtype preservation. Ideally would like to keep code here at a minimum

TomAugspurger replied:
Without it, you get a float:

```python
In [16]: s = pd.Series(dtype=int)

In [17]: s.add(t, fill_value=0)
Out[17]:
0    1.0
1    2.0
dtype: float64

```

I think it'd be strange for a `value_counts` to return floating-point values in the counts.

WillAyd reviewed (commented):
Minor quibbles but I'd be OK with it as is anyway

simonjayhawkins commented on environment.yml:
```diff
@@ -35,6 +35,8 @@ dependencies:
   - nbconvert>=5.4.1
   - nbsphinx
   - pandoc
+  - dask
```
could this be an optional dependency?

TomAugspurger replied:
This is for the dev env. We've been including all the dependencies necessary to build the docs.

simonjayhawkins replied:
must have been some misunderstanding in https://github.com/pandas-dev/pandas/pull/27646#issuecomment-518292531. That was a dev-only dependency.

jorisvandenbossche replied:
> This is for the dev env. We've been including all the dependencies necessary to build the docs.

That said, would it be possible to rely on just dask-core (+ what is needed for dask.dataframe), as distributed brings in a lot more dependencies? 
(the one code block that shows the client could be a code-block)

jorisvandenbossche commented on doc/source/user_guide/scale.rst:
```diff
@@ -0,0 +1,404 @@
+*************************
+
+Pandas provides data structures for in-memory analytics, which makes using pandas
+to analyze larger than memory datasets somewhat tricky.
```
This document is not only for "larger than memory" data right? It becomes already tricky if your dataset is (some factor) smaller than your memory, right? (because we create copies, because reading can take more memory, ...)

At least the first sections in this document equally apply as performance considerations on smaller-than-memory datasets

TomAugspurger replied:
Tried to clarify this a bit (in part by removing the "use efficient file formats" section.

jorisvandenbossche commented on doc/source/user_guide/scale.rst:
```diff
@@ -0,0 +1,404 @@
+   %time _ = pd.read_parquet("timeseries.parquet")
+
+Notice that parquet gives higher performance for reading (and writing), both
+in terms of speed and lower peak memory usage. See :ref:`io` for more.
```
Maybe link to the section in io.rst that compares the performance of different formats?

jorisvandenbossche commented on doc/source/user_guide/scale.rst:
```diff
@@ -0,0 +1,404 @@
+
+Some workloads can be achieved with chunking: splitting a large problem like "convert this
+directory of CSVs to parquet" into a bunch of small problems ("convert this individual parquet
+file into a CSV. Now repeat that for each file in this directory."). As long as each chunk
```
"convert this individual parquet file into a CSV" -> "convert this individual CSV file into a Parquet file" ?

(then in matches the example of the previous sentence)

jorisvandenbossche commented on doc/source/user_guide/scale.rst:
```diff
@@ -0,0 +1,404 @@
+
+Pandas is just one library offering a DataFrame API. Because of its popularity,
+pandas' API has become something of a standard that other libraries implement.
+The pandas documentation maintains a list of libraries implemetning a DataFrame API
```
```suggestion
The pandas documentation maintains a list of libraries implementing a DataFrame API
```

jorisvandenbossche commented on doc/source/user_guide/scale.rst:
```diff
@@ -0,0 +1,404 @@
+
+.. ipython:: python
+
+   divisions = tuple(pd.to_datetime(starts)) + (pd.Timestamp(ends[-1]),)
```
Those starts and ends were computed in a hidden cell ...

jorisvandenbossche reviewed (commented):
Nice addition to the docs!

Not for this PR, but if we have an additional level of navigation, we could put this together with the current enhancingperf.rst in a "performance" section?

TomAugspurger committed:
updates

TomAugspurger commented:
> Not for this PR, but if we have an additional level of navigation, we could put this together with the current enhancingperf.rst in a "performance" section?

I slightly prefer smaller doc pages that cross reference eachother.

I've removed the section comparing the perf of read_csv and read_parquet. That muddies the purpose of this document, since it's talking about speed, while the rest of the document focuses primarily on memory usage.

I've updated the "subset of columns" example to use the `%memit` magic from memory_profiler.

TomAugspurger committed:
fixup

TomAugspurger committed:
fixup

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into scale

TomAugspurger committed:
try updating check

TomAugspurger committed:
remove memory profiling

TomAugspurger committed:
fixup

TomAugspurger commented:
flake8-rst didn't like jupyter magics, so I've had to remove it.

jorisvandenbossche commented:
> > Not for this PR, but if we have an additional level of navigation, we could put this together with the current enhancingperf.rst in a "performance" section?
>
> I slightly prefer smaller doc pages that cross reference eachother.

Sorry, I was not very clear :) I didn't mean to suggest to put them in a single file. Because I also prefer smaller doc pages, I want to split some of the existing ones, and create an extra level of hierarchy in the user guide. And in that light, this one could fit together with the existing enhancingperf (which could be splitted for the cython stuff and the querying) in a general "Performance" section (that has multiple sub pages). But that's for the future, not now, so nothing to care about for this PR.

> I've removed the section comparing the perf of read_csv and read_parquet. That muddies the purpose of this document, since it's talking about speed, while the rest of the document focuses primarily on memory usage.

I actually think that section could still be useful. read_csv vs read_parquet is not only for speed, also memory wise I would think that read_parquet is quite a bit better? (I heard about read_csv needing quite some more memory than the original file or final dataframe) So maybe if using `%memit` instead of `%timeit` would illustrate that?

jorisvandenbossche commented on pandas/tests/groupby/test_categorical.py:
```diff
@@ -782,7 +782,8 @@ def test_categorical_no_compress():
 
-    # http://stackoverflow.com/questions/23814368/sorting-pandas-categorical-labels-after-groupby  # noqa: flake8
+    # http://stackoverflow.com/questions/23814368/
+    #   sorting-pandas-categorical-labels-after-groupby  # noqa: flake8
```
This is fixed on master (it's also what is causing the merge conflict)

jorisvandenbossche commented on environment.yml:
```diff
@@ -35,8 +35,11 @@ dependencies:
+  - toolz>=0.7.3
+  - fsspec>=0.5.1
+  - partd>=0.3.10
+  - cloudpickle>=0.2.1
```
Can you add a comment here that those last 4 ones are just dependencies of dask.dataframe?

jreback commented:
lgtm, needs a rebase though.

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into scale

TomAugspurger committed:
whatsnew, fixups

TomAugspurger commented:
All green here.

WillAyd reviewed (approved):
lgtm

TomAugspurger commented:
Thanks.

jorisvandenbossche commented:
No that it needed to be in this PR (could be a follow-up), but an opinion on the usefulness of what I remarked in https://github.com/pandas-dev/pandas/pull/28577#issuecomment-535467605 (the second paragraph on reading files)
